datafusion/optimizer/src/handle_gap_fill.rs

```diff
@@ -0,0 +1,451 @@
+                    new_group_expr,
+                    aggr_expr,
+                )?),
+            }),
+        })
+    };
+
+    Ok(Some(new_gap_fill_plan))
+}
+
+// Iterate over the group expression list.
+// If it finds no occurrences of date_bin_gapfill at the top of
+// each expression tree, it will return None.
+// If it finds such an occurrence, it will return a new expression list
+// with the date_bin_gapfill replaced with date_bin, and the index of
+// where the replacement occurred.
+fn replace_date_bin_gapfill(
+    group_expr: &Vec<Expr>,
+) -> Result<Option<(Vec<Expr>, usize)>> {
+    let has_date_bin_gapfill = group_expr.iter().any(|e| {
+        matches!(
+            e,
+            Expr::ScalarFunction {
+                fun: DateBinGapfill,
+                ..
+            }
+        )
+    });
+    if !has_date_bin_gapfill {
+        return Ok(None);
+    }
+
+    let mut new_aggr_expr = Vec::with_capacity(group_expr.len());
+    let mut dbg_idx = None;
+
+    group_expr
+        .iter()
+        .enumerate()
+        .try_for_each(|(i, e)| -> Result<()> {
+            let new_expr = match e {
+                Expr::ScalarFunction {
+                    fun: DateBinGapfill,
+                    args,
+                } => {
+                    if dbg_idx.is_some() {
+                        return Err(DataFusionError::Plan(format!(
+                            "DATE_BIN_GAPFILL specified more than once"
+                        )));
+                    }
+                    dbg_idx = Some(i);
+                    Expr::ScalarFunction {
+                        fun: DateBin,
+                        args: args.clone(),
+                    }
+                }
+                _ => e.clone(),
+            };
+            new_aggr_expr.push(new_expr);
+            Ok(())
+        })?;
+    Ok(Some((new_aggr_expr, dbg_idx.unwrap())))
+}
+
+fn collect_gap_fns(e: &Expr) -> Result<Vec<BuiltinScalarFunction>> {
+    struct Finder {
+        fns: Vec<BuiltinScalarFunction>,
+    }
+    impl ExpressionVisitor for Finder {
+        fn pre_visit(mut self, expr: &Expr) -> Result<Recursion<Self>> {
+            if let Expr::ScalarFunction {
+                fun:
+                    f @ (BuiltinScalarFunction::DateBinGapfill | BuiltinScalarFunction::LOCF),
+                ..
+            } = expr
+            {
+                self.fns.push(f.clone())
+            };
+            Ok(Recursion::Continue(self))
+        }
+    }
+    let f = Finder { fns: vec![] };
+    let f = e.accept(f)?;
+    Ok(f.fns)
+}
+
+fn check_node(node: &LogicalPlan) -> Result<()> {
+    println!("checking node");
+    node.expressions().iter().try_for_each(|expr| {
+        let fns = collect_gap_fns(expr)?;
+        if !fns.is_empty() {
+            Err(DataFusionError::Plan(format!(
+                "invalid context for {}",
+                fns[0]
+            )))
+        } else {
+            Ok(())
+        }
+    })?;
+    Ok(())
```

Comment:
I think you can just return the result directly here if you wanted:

```suggestion
    })
```